Deprecations and cleanup

TacoTheDank opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe. Please take a moment to fill out the following suggestion on how to structure this PR description. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [ ] Bug fix (user facing)
- [ ] Feature (user facing)
- [x] Code base improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
<!-- While bullet points are the norm in this section, feel free to write a text instead if you can't fit it in a list -->
- Fixes a bunch of deprecations
- A little bit of code cleanup

#### Fixes the following issue(s)
<!-- Also add reddit or other links which are relevant to your change. -->
- Fixes some deprecations in code

#### Testing apk
<!-- Ensure that you have your changes on a new branch which has a meaningful name. This name will be used as a suffix for the app ID to allow installing and testing multiple versions of NewPipe. Do NOT name your branches like "patch-0" and "feature-1". For example, if your PR implements a bug fix for comments, an appropriate branch name would be "commentfix". -->
[newpipe-deprecationsandcleanup.zip](https://github.com/TeamNewPipe/NewPipe/files/4963478/newpipe-deprecationsandcleanup.zip)

#### Agreement
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

wb9688 commented:
Code looks OK, though have you tested your changes on at least Android 4.4 and Android 10?

We also prefer marking variables as `final` whenever possible, e.g. in those enhanced for loops.

TacoTheDank commented:
@wb9688

I have no 4.4 device to test, and can't really do an AS emulator. I have an Android 10 one though, and will test on that.

I fixed a crash caused by one of the changes, so I reverted that (and better grouped the commits). I'm testing an APK, and everything works fine as far as I can tell.

MD77MD commented:
I have a 4.4 device.. what should I look for to test... could you put the link to the APK.

MD77MD commented:
I really think this will be more useful and needed after the big merge (expendable players with unified ui). with the release of newPipe 0.20.

TacoTheDank commented:
@MD77MD Didn't see this. I've added the APK. (I think it's the right one)

Stypox commented:
@MD77MD since you own a 4.4 device, are you willing to add yourself to the [testers page](https://github.com/TeamNewPipe/NewPipe/wiki/Testers) on the wiki?

Domiuns commented:
I have test the last apk on my phone android 4.4 and all work well. I continue with test if i have issues i post here, thanks for work :)

TobiGr commented:
Code looks good, but will be merged after #2907 as it changes player code.

TacoTheDank commented:
Ok. I'll probably have to fix conflicts but that's fine.

TacoTheDank commented:
Fixed. Some changes have been made as well, so re-review the commits.

Redirion commented:
@TacoTheDank please update the ExoPlayer-Dependency to 2.11.8 as 2.11.8 contains only "[trivial changes](https://github.com/google/ExoPlayer/commit/185be99d6137d782aa5fb572cb7ef127456b0d26)" it should be safe to do so

wb9688 commented on app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java:
```diff
@@ -106,7 +106,7 @@ public Socket createSocket(final InetAddress address, final int port,
 
     private Socket enableTLSOnSocket(final Socket socket) {
-        if (socket != null && (socket instanceof SSLSocket)) {
+        if ((socket instanceof SSLSocket)) {
```
No need for the double `((` and `))`

TacoTheDank replied:
Whoops.

wb9688 commented on app/src/main/java/org/schabi/newpipe/RouterActivity.java:
```diff
@@ -572,7 +568,7 @@ private String trim(final String input) {
             }
         }
-        return result.toArray(new String[result.size()]);
+        return result.toArray(new String[0]);
```
Why was this changed?

TacoTheDank replied:
![image](https://user-images.githubusercontent.com/32376686/91487190-5f8e3c80-e87b-11ea-8658-49bb02d561ca.png)

wb9688 replied:
Oh, OK

wb9688 commented on app/src/main/java/org/schabi/newpipe/player/resolver/VideoPlaybackResolver.java:
```diff
@@ -115,7 +115,7 @@ public MediaSource resolve(@NonNull final StreamInfo info) {
         } else {
             return new MergingMediaSource(mediaSources.toArray(
-                    new MediaSource[mediaSources.size()]));
+                    new MediaSource[0]));
```
Why was this changed?

TacoTheDank replied:
See above.

wb9688 reviewed (changes requested):
In general I like these changes. Could you rebase on `dev` and fix the Checkstyles errors though?

wb9688 commented:
You still haven't fixed all Checkstyle errors. Did you even try compiling the code before pushing it?

TacoTheDank commented:
@wb9688 
> You still haven't fixed all Checkstyle errors. Did you even try compiling the code before pushing it?

Yeah, I did. I just had to undo the commits and re-push them. It's much easier to find where the errors are via travis than trying to play find-the-error with this:
![image](https://user-images.githubusercontent.com/32376686/91490332-476cec00-e880-11ea-8c10-9651e37b5397.png)

Please watch the tone...

wb9688 commented:
@TacoTheDank: Ah, if it looks like that, it makes sense indeed, sorry. It shouldn't look like that though, are you using something other than Android Studio?

TacoTheDank commented:
@wb9688 Nope, it's just plain ol' Android Studio. The runCheckstyle just tells me the number of errors/warnings in the files, but doesn't give me anything else to work with :/

TacoTheDank commented:
I don't like how checkstyle operates. I've fixed all the errors, but it still errors because of the hashCode warning that I don't have anything to do with. I'll now have to fix that PR-unrelated thing for it to actually succeed :/

wb9688 commented:
@TacoTheDank: Oh, that's very strange. It should say e.g. the line number and that should be clickable to jump to the error. If you run `./gradlew runCheckstyle` in a terminal, does it show the errors and warnings with the line numbers (like on Travis)? Btw there are also some Ktlint errors now, but they could be automatically resolved by running `./gradlew formatKtlint`.

TacoTheDank commented:
I'll try it.

wb9688 commented:
@TacoTheDank: No, you don't have to fix the warnings. Notice how it fails at `:runKtlint` now ;)

TacoTheDank commented:
Ok, everything should NOW be good 😅

TacoTheDank commented:
Also, running `./gradlew runCheckstyle` now just builds successfully without giving me any warnings at all. Very strange lol

TacoTheDank commented:
I like my PRs being very clean, with no unnecessary after-changes. Sorry for all the force-pushing lol

opusforlife2 commented:
It's fine. I don't mind. You can force-push all you want. 🤭

MD77MD commented:
https://github.com/TeamNewPipe/NewPipe/pull/3909#issuecomment-682191436  that sounds naughty 😁

opusforlife2 commented:
Oh no no no. All force-pushing is 100% consensual in this repo, no matter how scary the term sounds. 🤣

Redirion commented:
> Code looks good, but will be merged after #2907 as it changes player code.

@TobiGr In my opinion this is good to merge now.

Stypox reviewed (approved):
Code looks good to me. Have you tested that everything works correctly on Android 4.4 with `SwitchPreferenceCompat`? (it should, but I remember strange shenanigans a while ago)

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/MainPlayer.java:
```diff
@@ -307,9 +307,7 @@ void resetNotification() {
-        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN) {
-            builder.setPriority(NotificationCompat.PRIORITY_MAX);
-        }
+        builder.setPriority(NotificationCompat.PRIORITY_MAX);
```
Heads up @friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java:
```diff
@@ -303,10 +302,6 @@ public static int getTossFlingVelocity(@NonNull final Context context) {
     public static CaptionStyleCompat getCaptionStyle(@NonNull final Context context) {
-        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
-            return CaptionStyleCompat.DEFAULT;
-        }
```
Heads up @friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java:
```diff
@@ -331,10 +326,6 @@ public static CaptionStyleCompat getCaptionStyle(@NonNull final Context context)
     public static float getCaptionScale(@NonNull final Context context) {
-        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
-            return 1f;
-        }
```
Heads up @friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/util/PermissionHelper.java:
```diff
@@ -26,10 +26,8 @@
-                return false;
-            }
+        if (!checkReadStoragePermissions(activity, requestCode)) {
+            return false;
```
@friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:
```diff
@@ -1105,7 +1105,7 @@ private void openBackgroundPlayer(final boolean append) {
         }
 
-        if (!useExternalAudioPlayer && android.os.Build.VERSION.SDK_INT >= 16) {
+        if (!useExternalAudioPlayer) {
```
Heads up @friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:
```diff
@@ -1302,16 +1302,9 @@ private void prepareDescription(final Description description) {
-                        }
-                        return parsedDescription;
-                    })
+                    .map((@NonNull String descriptionText) ->
+                            HtmlCompat.fromHtml(descriptionText,
+                                    HtmlCompat.FROM_HTML_MODE_LEGACY))
```
@friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java:
```diff
@@ -476,11 +474,8 @@ public void onPlaying() {
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-            playbackSeekBar.getThumb().setColorFilter(Color.RED, PorterDuff.Mode.SRC_IN);
-        }
+        playbackSeekBar.getThumb()
+                .setColorFilter(new PorterDuffColorFilter(Color.RED, PorterDuff.Mode.SRC_IN));
```
@friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java:
```diff
@@ -456,11 +457,8 @@ public void onBlocked() {
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-            playbackSeekBar.getThumb().setColorFilter(Color.RED, PorterDuff.Mode.SRC_IN);
-        }
+        playbackSeekBar.getThumb()
+                .setColorFilter(new PorterDuffColorFilter(Color.RED, PorterDuff.Mode.SRC_IN));
```
@friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java:
```diff
@@ -210,18 +211,18 @@ public void initViews(final View view) {
-        }
-        this.playbackSeekBar.getProgressDrawable().
-                setColorFilter(Color.RED, PorterDuff.Mode.MULTIPLY);
+        playbackSeekBar.getThumb()
+                .setColorFilter(new PorterDuffColorFilter(Color.RED, PorterDuff.Mode.SRC_IN));
```
@friendlyanon

TobiGr commented on app/src/main/java/org/schabi/newpipe/streams/Mp4DashReader.java:
```diff
@@ -259,11 +259,7 @@ private String boxName(final Box ref) {
-        } catch (final UnsupportedEncodingException e) {
-            return "0x" + Integer.toHexString(type);
-        }
+        return new String(ByteBuffer.allocate(4).putInt(type).array(), StandardCharsets.UTF_8);
```
@friendlyanon StandardCharsets was added with API 19

TobiGr reviewed (commented):
Looks good to me.
At the reviewers: please tell @friendlyanon whenever there are changes which might cause problems with the legacy version to not introduce preventable regressions. @friendlyanon You basically need to take a look at 1da206363eee8acd4ab2a2d325f84cdcdcacf31e and `StandardCharset` changes.

TacoTheDank commented:
@Stypox Unfortunately I do not have a KitKat device on which to test. :(

Stypox commented:
Ok, I will test myself on an emulator then 😉

Stypox commented:
I fixed a problem with the `TabAdapter` of `VideoDetailFragment`: for some reason the `BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT` had this issue: crash if enqueueing stream in the background and then clicking on it to open VideoDetailFragment: "Cannot setMaxLifecycle for Fragment not attached to FragmentManager". This can be solved by updating everything to [ViewPager2](https://developer.android.com/jetpack/androidx/releases/viewpager2) and removing `TabAdapter` altogether, but this is not relevant to this PR.
Everything else I checked worked fine, even on Android 4.4.

Stypox reviewed (approved):
Ready for merge in my opinion @wb9688. Thank you @TacoTheDank :-D

TacoTheDank commented:
I agree with migrating to ViewPager2.

Stypox commented:
@TacoTheDank thank you for your effort!
